readme.md
------

# YET ANOTHER YOUTUBE BBCODE TAG v4.14

[**By Dougiefresh**](http://www.simplemachines.org/community/index.php?action=profile;u=253913) -> [Link to Mod](http://custom.simplemachines.org/mods/index.php?mod=3982)

------

## Introduction
This modification adds a BBCode to the forum that shows a YouTube video in your post.

These BBCode takes these forms:
    
    [youtube]{youtube link}[/youtube]
    [youtube width=x height=y]{youtube link}[/youtube]
    [yt]{youtube ID}[/yt]
    [yt {parameters}]{youtube ID}[/yt]
    [yt_user]{username}[/yt_user]
    [yt_user {parameters}]{username}[/yt_user]
    [yt_search]{search query}[/yt_search]
    [yt_search {parameters}]{search query}[/yt_search]
    
where **width** and **height** is specified by **x** and **y**.  If width and height aren't specified, width is 100% of the post display area and height in a 16:9 ratio to the width.

The **yt_user** bbcode will show videos by the specified YouTube username.

The **yt_search** bbcode will show videos matching the provided search phrase.

This mod should be able to display a valid YouTube video when passing URLs.  For example, all of the following will display the YouTube video at [url]http://www.youtube.com/v/fA4cphzsjn8[/url].
[quote]
**For YouTube Videos:**
[nobbc]fA4cphzsjn8
http://www.youtube.com/fA4cphzsjn8
http://www.youtube.com/embed/fA4cphzsjn8
http://www.youtube.com/embed?v=fA4cphzsjn8
http://www.youtube.com/embed?feature=player_embedded&v=fA4cphzsjn8
http://www.youtube.com/watch?v=fA4cphzsjn8
http://www.youtube.com/v/fA4cphzsjn8
http://www.youtube.com/e/fA4cphzsjn8
http://www.youtube.com/?v=fA4cphzsjn8
http://www.youtube.com/user/username#p/u/11/fA4cphzsjn8
http://www.youtube.com/sandalsResorts#p/c/54B8C800269D7C1B/0/fA4cphzsjn8
http://www.youtube.com/watch?feature=player_embedded&v=fA4cphzsjn8
http://www.youtube.com/?feature=player_embedded&v=fA4cphzsjn8[/nobbc]

**For YouTube PlayLists:**
[nobbc]PL55713C70BA91BD6E
PLquckZj9TVRFKWqM6LJI4a_dgvN48jZGk
http://www.youtube.com/e/PL55713C70BA91BD6E
http://www.youtube.com/p/PL55713C70BA91BD6E
http://www.youtube.com/embed/PL55713C70BA91BD6E
http://www.youtube.com/embed?list=PL55713C70BA91BD6E
http://www.youtube.com/embed?feature=player_embedded&list=PL55713C70BA91BD6E
http://www.youtube.com/watch?v=OPf0YbXqDm0&list=PL55713C70BA91BD6E
http://www.youtube.com/?list=PL55713C70BA91BD6E
http://www.youtube.com/?feature=player_embedded&list=PL55713C70BA91BD6E
https://www.youtube.com/embed/videoseries?list=PLquckZj9TVRFKWqM6LJI4a_dgvN48jZGk
https://www.youtube.com/watch?v=Ni4ZclaiOtY&list=PLquckZj9TVRFKWqM6LJI4a_dgvN48jZGk[/nobbc]
[/quote]
Substituting **http://youtube.com**, **http://www.youtube-nocookies.com**, **http://youtu.be**, and **http://www.youtu.be** also works.  **https://** works instead of **http://**.  Please note any other parameters specified in the URL are ignored.

Optional parameters and values for all YouTube bbcodes are:

- **width** = Specifies the width of the iframe area.
- **height** = Specifies the height of the iframe area.
- **start** = Specifies how far into the video to start at.  Valid formats: "Seconds" or "Minutes:Seconds"
- **end** = Specifies how far into the video to stop at.  Valid formats: "Seconds" or "Minutes:Seconds"
- **autoplay** = **1**, **yes**, **on** or **true**
- **color** = **red** or **white**
- **theme** = **dark** or **light**
- **loop** = **1**, **yes**, **on** or **true**
- **controls** = **0**, **no**, **off**, **hide** or **false**
- **showinfo** = **0**, **no**, **off**, **hide** or **false**
- **privacy** = **1**, **yes**, **on** or **true**

## Profile Settings
There is a new option under **Profile** => **Look and Layout** called **Show YouTube videos as a link**.  Checking this box makes the mod create links instead of embedded videos in your post.

## Admin Settings
This BBCode may be disabled by going into the **Admin** => **Forum** => **Posts and Topics** => **Bulletin Board Code** and unchecking the bbcodes you don't want to use.  You may also be uninstall this mod in order to disable it.

## Related Discussions

- [Adding option to stop mod from displaying YouTube video's as a video](http://www.simplemachines.org/community/index.php?topic=522345.msg3695157#msg3695157)
- [Youtube I.D parsing for new URL formats](http://stackoverflow.com/questions/2936467/parse-youtube-video-id-using-preg-match)
- [Need help with a Regular Expression for YouTube links....](http://www.simplemachines.org/community/index.php?topic=533748.0)

## Extra Credits Go Out To:

- [Sapozhnik](http://www.simplemachines.org/community/index.php?action=profile;u=317117) for code assistance
- [karavan2](http://www.simplemachines.org/community/index.php?action=profile;u=308325) for providing code assistance
- [kelvincool](http://www.simplemachines.org/community/index.php?action=profile;u=303481) for Regular Expression assistance

## Translators

- Spanish Latin: [Rock Lee [BC]](http://www.simplemachines.org/community/index.php?action=profile;u=322597)

## Compatibility Notes
This mod was tested on SMF 2.0.10, but should work on SMF 2.1 Beta 3, as well as SMF 2.0 and up.  SMF 2.1 Beta 2 and SMF 1.x will not be supported.

## Changelog
The changelog can be viewed at [XPtsp.com](http://www.xptsp.com/board/free-modifications/yet-another-youtube-bbcode/?tab=1).

## License
Copyright (c) 2015 - 2018, Douglas Orend

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changelog.txt
v4.14 - November 9th, 2018
o No functionality change.
o Updated documentation to point to new website.

v4.13 - April 21st, 2018
o Updated Spanish Latin translation, courtsey of [url=http://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee[/url].

v4.12 - April 8th, 2018
o Fixed issue with ADK Portal that prevented video from being shown at proper size.

v4.11 - March 8th, 2018
o Added missing option to disable YouTube bbcodes in signatures.
o Added new option to disable autoembedding for YouTube links.

v4.10 - March 4th, 2018
o Fixed issue where BBC manager didn't show any tags.

v4.9 - February 22nd, 2018
o Updated Spanish Latin translation, courtsey of [url=http://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee[/url].

v4.8 - February 20th, 2018
o Removed support for SMF 2.1 Beta 2.
o Added support for SMF 2.1 Beta 3.
o Moved bbcode disabling code from [b]Sources/Subs.php[/b] to [b]Subs-BBCode-YouTube.php[/b].
o Added translators to the mod's [b]README.txt[/b] file.
o Added [b]LICENSE.txt[/b] to satisfy licensing concerns.

v4.7 - October 17th, 2017
o Removed option to force HTTPS mode.
o Modified code to force HTTPS mode in all YouTube tags.

v4.6 - October 9th, 2017
o Added option to force using HTTPS instead of allowing HTTP in links.

v4.5 - August 31st, 2017
o Fixed undefined array element error in [b]Subs-BBCode-YouTube.php[/b].

v4.4 - August 24th, 2017
o Fixed auto-embed so that [b]youtu.be/RkZkekS8NQU[/b] will embed itself properly.
o Added option to allow to embed YouTube videos in signatures. (BAD IDEA!!)

v4.3 - July 16th, 2017
o Added ability for user to full-screen the video.

v4.2 - July 6th, 2017
o Reorganized file edits to support SMF 2.1 Beta 3 properly.
o Removed support for SMF 2.1 Beta 2.

v4.1 - October 24th, 2016
o Added code to SimplePortal code to skip embedding links for shoutboxes.

v4.0 - September 14th, 2016
o Added buttons for the [b]yt_user[/b] and [b]yt_search[/b] bbcodes.
o Added minute:second validation to [b]start[/b] and [b]end[/b] parameters.

v3.11 - April 7th, 2016
o Added Spanish Latin translation by [url=http://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee [BC][/url].

v3.10 - March 7th, 2016
o Modified auto-embed code so that YouTube URLs in signature are embedded as [b][nobbc][url]{link}[/url][/nobbc][/b].
o Merged English and English UTF8 language files together.

v3.9 - January 24th, 2016
o Added code to make YouTube bbcode recognize content as 34-character playlist.
o Added code to force aspect ratio of width to height to be 16:9.
o Modified CSS file so that there isn't a 25px gap at the top of the video.

v3.8 - January 23rd, 2016
o Modified new YouTube video playlist URLs support for better coverage.

v3.7 - January 23rd, 2015
o Added support for the new YouTube video playlist URLs.
o Changed regular expression to make sure "?list=", "&list=", "?v=" and "&v=" are found first.

v3.6 - November 30th, 2015
o Added auto-embed for YouTube URLs.

v3.5 - November 4th, 2015
o SMF 2.1: Fixed BBCode-YouTube.css install location.

v3.4 - August 21st, 2015
o Fixed issue where necessary tag cannot be added if other mods have modified same lines.
o Rewrote the "fix parameter order" function to be called only when necessary.
o Operation that calls "fix parameter order" function now skipped for nightly SMF 2.1 Beta 2.

v3.3 - August 12th, 2015
o Added settings to allow admin to set default size of the YouTube video.
o Altered code to use new default size settings when width and/or height isn't specified.

v3.2 - August 8th, 2015
o Removed support for SMF 2.1 Beta 1
o Added support for SMF 2.1 Beta 2

v3.1 - August 1st, 2015
o Performance increase due to modification of parse_bbc function to deal with memory issue found.

v3.0 - July 17th, 2015
o CrescentEclipse added responsive video display elements to this mod.  Thanks, CrescentEclipse!
o When specifying width and height parameters, they are now treated as maximum display size.
o Changed default width and height from 640x400 to 100% of post display area.
o Modified "start" and "end" parameters to use "minutes:seconds" format in addition to "seconds".

v2.5 - April 13th, 2015
o Fixed mistake introduced in v2.4 regarding YouTube links.

v2.4 - April 12th, 2015
o Fixed YouTube URLs used for mobile devices.

v2.3 - May 1st, 2015
o Fixed missing BBCode button in editor for SMF 2.1 Beta 1.
o Fixed a bug where using YouTube ID instead of URL showed error message.

v2.2 - April 6th, 2015
o Updated for SMF 2.1 Beta 1

v2.1 - February 19th, 2015
o Readded code segment from v1.8 that dealt with "method of last resort" video ID.

v2.0 - February 19th, 2015
o YouTube videos are now displayed using the iframe tag for HTML5 compatibility.
o http://www.youtube-nocookie.com is now a valid URL to specify in the bbcode.
o Added yt_user bbcode to create a playlist of YouTube videos by a specific user.
o Added yt_search bbcode to create a playlist based on specified search results.
o Added support for YouTube playlist IDs as part of the youtube and yt bbcodes.
o Modification made to BBCode hook to define all bbcode parameters only ONCE for uniformity.
o Fixed a bug in the validation function that didn't set width/height parameters correctly.
o Modified validation function to work properly for all known YouTube video/playlist URLs with IDs.
o Parameters width and height are now optional instead of mandatory.
o Fixed an issue where the video's aspect ratio is screwed up if width or height isn't specified.
o Added autoplay parameter to the bbcodes.  1, yes, on and true are supported.
o Added color parameter to the bbcodes.  red and white are supported by YouTube.
o Added theme parameter to the bbcodes.  dark and light are supported by YouTube.
o Added loop parameter to the bbcodes.  1, yes, on and true are supported.
o Added start parameter to the bbcodes, specifying the number of seconds into the video to start at.

add_remove_hooks.php
<?php
/**********************************************************************************
* add_remove_hooks.php                                                            *
***********************************************************************************
***********************************************************************************
* This program is distributed in the hope that it is and will be useful, but      *
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY    *
* or FITNESS FOR A PARTICULAR PURPOSE.                                            *
*                                                                                 *
* This file is a simplified database installer. It does what it is suppoed to.    *
**********************************************************************************/
global $forum_version;

// If we have found SSI.php and we are outside of SMF, then we are running standalone.
if (file_exists(dirname(__FILE__) . '/SSI.php') && !defined('SMF'))
	require_once(dirname(__FILE__) . '/SSI.php');
elseif (!defined('SMF')) // If we are outside SMF and can't find SSI.php, then throw an error
	die('<b>Error:</b> Cannot install - please verify you put this file in the same place as SMF\'s SSI.php.');
if (SMF == 'SSI')
	db_extend('packages');
	
// Define the hooks
$hook_functions = array(
	'integrate_pre_include' => '$sourcedir/Subs-BBCode-YouTube.php',
	'integrate_load_theme' => 'BBCode_YouTube_LoadTheme',
	'integrate_bbc_codes' => 'BBCode_YouTube',
	'integrate_bbc_buttons' => 'BBCode_YouTube_Button',
	'integrate_general_mod_settings' => 'BBCode_YouTube_Settings',
// SMF 2.1+ Hooks below this line:
	'integrate_pre_parsebbc' => 'BBCode_YouTube_Embed',
);

// Adding or removing them?
if (!empty($context['uninstalling']))
	$call = 'remove_integration_function';
else
	$call = 'add_integration_function';

// Do the deed
foreach ($hook_functions as $hook => $function)
	$call($hook, $function);

if (SMF == 'SSI')
   echo 'Congratulations! You have successfully installed this mod!';

?>

Subs-BBCode-YouTube.php
<?php
/**********************************************************************************
* Subs-BBCode-YouTube.php
***********************************************************************************
* This mod is licensed under the 2-clause BSD License, which can be found here:
*	http://opensource.org/licenses/BSD-2-Clause
***********************************************************************************
* This program is distributed in the hope that it is and will be useful, but      *
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY    *
* or FITNESS FOR A PARTICULAR PURPOSE.                                            *
**********************************************************************************/
if (!defined('SMF')) 
	die('Hacking attempt...');

//=================================================================================
// BBCode Hook functions
//=================================================================================
function BBCode_YouTube_Settings(&$config_vars)
{
	$config_vars[] = array('int', 'youtube_default_width');
	$config_vars[] = array('int', 'youtube_default_height');
	$config_vars[] = array('check', 'youtube_no_embed');
	$config_vars[] = array('check', 'youtube_disable_autoembed');
}

function BBCode_YouTube_LoadTheme()
{
	global $context, $settings, $sourcedir;
	
	// Add necessary headers and stuff for the forum:
	$context['html_headers'] .= '
	<link rel="stylesheet" type="text/css" href="' . $settings['default_theme_url'] . '/css/BBCode-YouTube2.css" />';
	$context['allowed_html_tags'][] = '<iframe>';
	
	// If ADK Portal is installed and we're using main page(s), add the integration hook:
	if (file_exists($sourcedir . '/AdkPortal/AdkPortal.php') && empty($_GET['action']))
		add_integration_function('integrate_buffer', 'BBCode_YouTube_Buffer', false);
}

function BBCode_YouTube_parameters()
{
	return array(
		'width' => array('optional' => true, 'match' => '(\d+)', 'validate' => 'BBCode_YouTube_width'),
		'height' => array('optional' => true, 'match' => '(\d+)', 'validate' => 'BBCode_YouTube_height'),
		'autoplay' => array('optional' => true, 'match' => '(1|yes|on|true)', 'validate' => 'BBCode_YouTube_autoplay'),
		'color' => array('optional' => true, 'match' => '(red|white)', 'validate' => 'BBCode_YouTube_color'),
		'theme' => array('optional' => true, 'match' => '(light|dark)', 'validate' => 'BBCode_YouTube_theme'),
		'loop' => array('optional' => true, 'match' => '(1|yes|on|true)', 'validate' => 'BBCode_YouTube_loop'),
		'start' => array('optional' => true, 'match' => '(\d+|\d+\:\d+)', 'validate' => 'BBCode_YouTube_start'),
		'end' => array('optional' => true, 'match' => '(\d+|\d+\:\d+)', 'validate' => 'BBCode_YouTube_end'),
		'privacy' => array('optional' => true, 'match' => '(1|yes|on|true)', 'validate' => 'BBCode_YouTube_privacy'),
		'controls' => array('optional' => true, 'match' => '(0|no|off|false|hide)', 'validate' => 'BBCode_YouTube_controls'),
		'showinfo' => array('optional' => true, 'match' => '(0|no|off|false|hide)', 'validate' => 'BBCode_YouTube_showinfo'),
	);
}

function BBCode_YouTube(&$bbc)
{
	// Since all the bbcodes use the same parameters, let's define them ONCE!
	$params = BBCode_YouTube_parameters();

	// Syntax: [youtube {params}]{URL}[/youtube]
	$bbc[] = array(
		'tag' => 'youtube',
		'type' => 'unparsed_content',
		'content' => '$1',
		'parameters' => $params,
		'validate' => 'BBCode_YouTube_URL',
		'disabled_content' => '$1',
	);

	// Syntax: [youtube]{URL}[/youtube]
	$bbc[] = array(
		'tag' => 'youtube',
		'type' => 'unparsed_content',
		'content' => '$1',
		'validate' => 'BBCode_YouTube_URL',
		'disabled_content' => '$1',
	);

	// Syntax: [yt {params}]{URL}[/yt]
	$bbc[] = array(
		'tag' => 'yt',
		'type' => 'unparsed_content',
		'content' => '$1',
		'parameters' => $params,
		'validate' => 'BBCode_YouTube_URL',
		'disabled_content' => '$1',
	);

	// Syntax: [yt]{URL}[/yt]
	$bbc[] = array(
		'tag' => 'yt',
		'type' => 'unparsed_content',
		'content' => '$1',
		'validate' => 'BBCode_YouTube_URL',
		'disabled_content' => '$1',
	);

	// Syntax: [yt_user {params}]{user name}[/yt_user]
	$bbc[] = array(
		'tag' => 'yt_user',
		'type' => 'unparsed_content',
		'content' => '$1',
		'parameters' => $params,
		'validate' => 'BBCode_YouTube_User',
		'disabled_content' => '$1',
	);

	// Syntax: [yt_user]{user name}[/yt_user]
	$bbc[] = array(
		'tag' => 'yt_user',
		'type' => 'unparsed_content',
		'content' => '$1',
		'validate' => 'BBCode_YouTube_User',
		'disabled_content' => '$1',
	);

	// Syntax: [yt_search {params}]{search spec}[/yt_search]
	$bbc[] = array(
		'tag' => 'yt_search',
		'type' => 'unparsed_content',
		'content' => '$1',
		'parameters' => $params,
		'validate' => 'BBCode_YouTube_Search',
		'disabled_content' => '$1',
	);

	// Syntax: [yt_search]{search spec}[/yt_search]
	$bbc[] = array(
		'tag' => 'yt_search',
		'type' => 'unparsed_content',
		'content' => '$1',
		'validate' => 'BBCode_YouTube_Search',
		'disabled_content' => '$1',
	);
}

function BBCode_YouTube_Button(&$buttons)
{
	global $txt;

	$buttons[count($buttons) - 1][] = array(
		'image' => 'youtube',
		'code' => 'youtube',
		'description' => $txt['youtube'],
		'before' => '[youtube]',
		'after' => '[/youtube]',
	);
	$buttons[count($buttons) - 1][] = array(
		'image' => 'yt_search',
		'code' => 'yt_search',
		'description' => $txt['yt_search'],
		'before' => '[yt_search]',
		'after' => '[/yt_search]',
	);
	$buttons[count($buttons) - 1][] = array(
		'image' => 'yt_user',
		'code' => 'yt_user',
		'description' => $txt['yt_user'],
		'before' => '[yt_user]',
		'after' => '[/yt_user]',
	);
}

//=================================================================================
// Parameter validation functions
//=================================================================================
function BBCode_YouTube_height($height)
{
	global $context;
	$context['bbc_youtube']['height'] = max(0, (int) $height);
}

function BBCode_YouTube_width($width)
{
	global $context;
	$context['bbc_youtube']['width'] = max(0, (int) $width);
}

function BBCode_YouTube_autoplay($autoplay)
{
	global $context;
	$context['bbc_youtube']['autoplay'] = 1;
}

function BBCode_YouTube_color($color)
{
	global $context;
	$context['bbc_youtube']['color'] = $color;
}

function BBCode_YouTube_theme($theme)
{
	global $context;
	$context['bbc_youtube']['theme'] = $theme;
}

function BBCode_YouTube_loop($loop)
{
	global $context;
	$context['bbc_youtube']['loop'] = 1;
}

function BBCode_YouTube_start($start)
{
	global $context;
	$minutes = 0;
	if (strpos($start, ':'))
	{
		list($minutes, $start) = explode(':', $start);
		$start = max(0, min(59, $start));
	}
	$context['bbc_youtube']['start'] = ($minutes * 60) + $start;
}

function BBCode_YouTube_end($end)
{
	global $context;
	$minutes = 0;
	if (strpos($end, ':'))
	{
		list($minutes, $end) = explode(':', $end);
		$end = max(0, min(59, $end));
	}
	$context['bbc_youtube']['end'] = ($minutes * 60) + $end;
}

function BBCode_YouTube_privacy($privacy)
{
	global $context;
	$context['bbc_youtube']['privacy'] = 1;
}

function BBCode_YouTube_controls($controls)
{
	global $context;
	$context['bbc_youtube']['controls'] = 1;
}

function BBCode_YouTube_showinfo($showinfo)
{
	global $context;
	$context['bbc_youtube']['showinfo'] = 1;
}

//=================================================================================
// YouTube and YT BBCode validation functions:
//=================================================================================
function BBCode_YouTube_URL(&$tag, &$data, &$disabled)
{
	global $txt, $context, $modSettings;

	// Do some preperation to avoid some issues within the code:
	$data = str_replace('&amp;', '&', strip_tags($data));

	// Start building the YouTube URL that we are going to show to the user:
	if (strpos($data, 'youtube-nocookie.com') !== false)
		$context['bbc_youtube']['privacy'] = 1;
	$server = 'https://www.youtube' . (isset($context['bbc_youtube']['privacy']) ? '-nocookie' : '') . '.com';

	// Figure out if what's been passed is a YouTube video URL or ID:
	if (($len = strlen($data)) == 11)
		$data = $server . '/embed/' . ($url = $data);
	elseif ($len == 18)
		$data = $server . '/embed?listType=playlist&list=' . ($url = $data);
	elseif ($len == 34 && substr($data, 0, 2) == 'PL')
		$data = $server . '/embed?listType=playlist&list=' . ($url = $data);
	else
		$data = $server . '/' . ($url = BBCode_YouTube_Parse($data));

	// If the URL variable is empty, return link invalid to user....
	if (empty($url))
		$data = $txt['youtube_link_invalid'];
	// If the YouTube bbcode is disabled, create a simple link to the video:
	elseif (isset($disabled['youtube']))
		$data = '<a href=' . $data . '>' . $data . '</a>';
	// Otherwise, build the YouTube HTML string that we're going to use:
	else
		BBCode_YouTube_Link($data);

	// Remove the validation variables from the context array:
	unset($context['bbc_youtube']);
}

function BBCode_YouTube_User(&$tag, &$data, &$disabled)
{
	global $txt, $context, $modSettings;

	// Create the actual URL we are going to be using:
	$server = 'https://www.youtube' . (isset($context['bbc_youtube']['privacy']) ? '-nocookie' : '') . '.com';
	$data = $server . '/embed?listType=user_uploads&list=' . str_replace(" ", "+", strip_tags($data));

	// If the YouTube bbcode is disabled, create a simple link to the video:
	if (isset($disabled['youtube']))
		$data = '<a href=' . $data . '>' . $data . '</a>';
	// Otherwise, build the YouTube HTML string that we're going to use:
	else
		BBCode_YouTube_Link($data);

	// Remove the validation variables from the context array:
	unset($context['bbc_youtube']);
}

function BBCode_YouTube_Search(&$tag, &$data, &$disabled)
{
	global $txt, $context, $modSettings;

	// Create the actual URL we are going to be using:
	$server = 'https://www.youtube' . (isset($context['bbc_youtube']['privacy']) ? '-nocookie' : '') . '.com';
	$data = $server . '/embed?listType=search&list=' . str_replace(" ", "+", strip_tags($data));

	// If the YouTube bbcode is disabled, create a simple link to the video:
	if (isset($disabled['youtube']))
		$data = '<a href=' . $data . '>' . $data . '</a>';
	// Otherwise, build the YouTube HTML string that we're going to use:
	else
		BBCode_YouTube_Link($data);

	// Remove the validation variables from the context array:
	unset($context['bbc_youtube']);
}

//=================================================================================
// YouTube and YT BBCode link creation function:
//=================================================================================
function BBCode_YouTube_Link(&$data)
{
	global $context, $modSettings;
	
	// If either width or height is not defined, then pull the default values:
	$width = (isset($context['bbc_youtube']['width']) ? $context['bbc_youtube']['width'] : (!empty($modSettings['youtube_default_width']) ? $modSettings['youtube_default_width'] : 0));
	$height = (isset($context['bbc_youtube']['height']) ? $context['bbc_youtube']['height'] : (!empty($modSettings['youtube_default_height']) ? $modSettings['youtube_default_height'] : 0));

	// Adjust the video width & height to match a 16:9 ratio:
	if (!empty($width) && !empty($height) && ($width / $height) != 0.5625)
		$height = floor($width * 0.5625);

	// Process the rest of the parameters we've saved in the validation functions:
	if (isset($context['bbc_youtube']['autoplay']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'autoplay=1';
	if (isset($context['bbc_youtube']['color']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'color=' . $context['bbc_youtube']['color'];
	if (isset($context['bbc_youtube']['theme']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'theme=' . $context['bbc_youtube']['theme'];
	if (isset($context['bbc_youtube']['start']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'start=' . $context['bbc_youtube']['start'];
	if (isset($context['bbc_youtube']['end']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'end=' . $context['bbc_youtube']['end'];
	if (isset($context['bbc_youtube']['loop']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'loop=1';
	if (isset($context['bbc_youtube']['controls']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'controls=0';
	if (isset($context['bbc_youtube']['showinfo']))
		$data = $data . (strpos($data, '?') !== false ? '&' : '?') . 'showinfo=0';

	// Build the HTML string that we are going to display to the user:
	$data = 
		'<div style="' . (!empty($width) ? 'max-width: ' . $width . 'px;' : '') . (!empty($height) ? ' max-height: ' . $height . 'px;' : '') . '">' .
			'<div class="yt-wrapper">' .
				'<iframe type="text/html" src="' . $data . '" allowfullscreen frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>' .
			'</div>' . 
		'</div>';
}

//=================================================================================
// Regular Expression function for URL validation:
//=================================================================================
function BBCode_YouTube_Parse($url)
{
	if (preg_match('#(\?list=|\&amp;list=|/|/e/|/embed/|/p/)(PL[\w-]{32}|[\w-]{18})#i', $url, $matches))
	{
		parse_str(parse_url(str_replace('&amp;', '&', $url), PHP_URL_QUERY), $out);
		return ('embed?listType=playlist&list=' . $matches[2] . (isset($out['v']) ? '&v=' . $out['v'] : '') . (isset($out['index']) ? '&index=' . $out['index'] : ''));
	}
	preg_match_all('#(\?v=|\&anp;v=|/|/v/|/e/|/embed/)([\w-]{11})#i', $url, $matches);
	return (isset($matches[2][count($matches[2]) - 1]) ? 'embed/' . $matches[2][count($matches[2]) - 1] : false);
}

//=================================================================================
// Function to fix parameter order of the YouTube bbcode:
//=================================================================================
function BBCode_YouTube_Params(&$message, $pos, &$parameters)
{
	$match = substr($message, $pos - 1);
	$match = substr($match, 0, strpos($match, ']'));
	$params = explode(' ', $match);
	unset($params[0]);
	$order = array();
	$replace_str = $old = '';
	foreach ($params as $param)
	{
		if (strpos($param, '=') === false && isset($order[$old]))
			$order[$old] .= ' ' . $param;
		else
			$order[$old = substr($param, 0, strpos($param, '='))] = substr($param, strpos($param, '=') + 1);
	}
	foreach ($parameters as $key => $ignore)
	{
		$replace_str .= (isset($order[$key]) ? ' ' . $key . '=' . $order[$key] : '');
		unset($order[$key]);
	}
	$message = str_replace($match, $replace_str, $message);
	return count($order) == 0;
}

//=================================================================================
// Function to auto-embed YouTube URLs:
//=================================================================================
function BBCode_YouTube_Embed(&$message, &$smileys, &$cache_id, &$parse_tags)
{
	global $options, $context, $modSettings;
	
	// Added to allow the user the option to not embed a YouTube video
	if (!empty($options['youtube_no_embed']))
		$modSettings['disabledBBC'] = (!empty($modSettings['disabledBBC']) ? $modSettings['disabledBBC'] . ',' : '') . 'youtube,yt,yt_search,yt_user';

	// Is message empty, OR autoembedding links disabled?  If so, return to caller:
	if ($message === false || !empty($context['bbc_youtube']['ignore']) || !empty($modSettings['youtube_disable_autoembed']))
		return;
	$replace = (strpos($cache_id, 'sig') !== false ? '[url]$0[/url]' : '[youtube]$0[/youtube]');
	$pattern = '~(?<=[\s>\.(;\'"]|^)(?:https?\:\/\/)(?:www\.)?(youtube\.com|youtube-nocookie\.com|youtu\.be)/?(?:/[\w\-_\~%\.@!,\?&;=#(){}+:\'\\\\]*)*(\?list=|\&amp;list=|/e/|/embed/|/p/)(PL[\w-]{32}|[\w-]{18})\??[/\w\-_\~%@\?;=#}\\\\]?~';
	$message = preg_replace($pattern, $replace, $message);
	$pattern = '~(?<=[\s>\.(;\'"]|^)(?:https?\:\/\/)(?:www\.)?(youtube\.com|youtube-nocookie\.com|youtu\.be)/?(?:/[\w\-_\~%\.@!,\?&;=#(){}+:\'\\\\]*)*(\?v=|\&amp;v=|/v/|/e/|/embed/|)([\w-]{11})(?:\&amp;list=(PL[\w-]{32}|[\w-]{18}))?\??[/\w\-_\~%@\?;=#}\\\\]?~';
	$message = preg_replace($pattern, $replace, $message);
	if (strpos($cache_id, 'sig') !== false && empty($modSettings['youtube_sig_embed']))
		$message = preg_replace('#\[youtube.*\](.*)\[\/youtube\]#i', '[url]$1[/url]', $message);
}

//=================================================================================
// Function to remove unnecessary class addition from YouTube videos:
//=================================================================================
function BBCode_YouTube_Buffer($buffer)
{
	return str_replace('class="resize_auto_new youtube-player"', 'class="youtube-player"', $buffer);
}

?>